From 8785f5d3ed881ee0eb6691332566a47c12e6d57a Mon Sep 17 00:00:00 2001
From: Sivashankari Madhavan <quic_sivamadh@quicinc.com>
Date: Tue, 13 Sep 2022 17:08:08 +0530
Subject: [PATCH] mac80211: Fix Green AP Crash issue

While execute the Green AP set command, filling the bss info changed flags
for AP_PS based on hw_param supports_ap_ps. In ath11k, supports_ap_ps is
not enabled. Due to this not filling the bss changed flags for AP_PS,
and processing the default BSS changed value -EINVAL (-22). Based on
this value some unexpected BSS_CHANGED Flags getting processed and
crash issue observed.

Added the pre-condition checking. If the err is less than of 0, then avoided
to update the BSS info changed Value.

Signed-off-by: Sivashankari Madhavan <quic_sivamadh@quicinc.com>
---
 net/mac80211/cfg.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1639,9 +1639,12 @@ static int ieee80211_update_ap(struct wi
 	if (params->beacon.he_bss_color_valid &&
 	    params->beacon.he_bss_color.enabled != link_conf->he_bss_color.enabled) {
 		link_conf->he_bss_color.enabled = params->beacon.he_bss_color.enabled;
-		changed |= BSS_CHANGED_HE_BSS_COLOR;
+		err |= BSS_CHANGED_HE_BSS_COLOR;
 	}
 
+	if (err > 0)
+		changed = err;
+
 	if (params->fils_discovery.max_interval) {
 		err = ieee80211_set_fils_discovery(sdata,
 						    &params->fils_discovery,
